api-server/src/main/java/io/enmasse/api/server/RequestLogger.java

```diff
@@ -70,19 +61,21 @@ public void filter(ContainerRequestContext containerRequestContext, ContainerRes
     }
 
     private void updateMetrics(int status) {
-        if (status >= 400 && status < 500) {
-            num4xx++;
-        } else if (status >= 500) {
-            num5xx++;
+        if (status < 0 || status >= MAX_CODE) {
+            return;
         }
-        numTotal++;
+        statusCodeCounter[status]++;
 
         Instant now = Instant.now(clock);
         if (now.minus(metricUpdateInterval).isAfter(lastMetricUpdate)) {
             long nowMillis = now.toEpochMilli();
-            totalRequestsMetric.updateValues(new MetricValue(numTotal, nowMillis));
-            code4xxRequestsMetrics.updateValues(new MetricValue(num4xx, nowMillis));
-            code5xxRequestsMetrics.updateValues(new MetricValue(num5xx, nowMillis));
+            List<MetricValue> values = new ArrayList<>();
+            for (int i = 0; i < MAX_CODE; i++) {
+                if (statusCodeCounter[i] > 0) {
+                    values.add(new MetricValue(statusCodeCounter[i], nowMillis, new MetricLabel("status", String.format("%3d", i))));
+                }
+            }
+            requestsMetric.updateValues(values);
```

Comment:
the synchronisation on this method will effectively single thread the API server every time `now.minus(metricUpdateInterval).isAfter(lastMetricUpdate)` is true.  There could be many threads awaiting this mutex.  Similarly, this is no guarantee that blocked threads will wake up in order so lastMetricUpdate may be updated incorrectly.